python/taichi/types/ndarray_type.py

```diff
@@ -9,78 +12,107 @@ def __init__(self, element_type, shape=None):
         self.layout = Layout.AOS
 
 
+# TODO(Haidong): This is a helper function that creates a MatrixType
+#                with respect to element_dim and element_shape.
+#                Remove this function when the two args are totally deprecated.
+def make_matrix_dtype_from_element_shape(element_dim, element_shape,
+                                         primitive_dtype):
+    if isinstance(primitive_dtype, CompoundType):
+        raise TypeError(
+            f'Cannot specifiy matrix dtype "{primitive_dtype}" and element shape or dim at the same time.'
+        )
+
+    # Scalars
+    if element_dim == 0 or (element_shape is not None
+                            and len(element_shape) == 0):
+        return primitive_dtype
+
+    # Cook element dim and shape into matrix type.
+    mat_dtype = None
+    if element_dim is not None:
+        if element_dim < 0 or element_dim > 2:
+            raise ValueError(
+                "Only scalars, vectors, and matrices are allowed as elements of ti.types.ndarray()"
+            )
+        # Check dim consistency. The matrix dtype will be cooked later.
+        if element_shape is not None and len(element_shape) != element_dim:
+            raise ValueError(
+                f"Both element_shape and element_dim are specified, but shape doesn't match specified dim: {len(element_shape)}!={element_dim}"
+            )
+        mat_dtype = vector(None,
+                           primitive_dtype) if element_dim == 1 else matrix(
+                               None, None, primitive_dtype)
+    if element_shape is not None:
```

Comment:
These `if`s are error prone in the future that it may slip away and return `mat_dtype` as None. 
Can we refactor this function to: 
```
if element_dim is not None:
  ...
  return ...
elif element_shape is not None:
  ...
  return ...
else:
  throw!
```

Reply:
I used to place some return statement at each branch, but forgot to change that if back to elif when I unified the returns. Thanks for catching this!